x-pack/plugins/uptime/public/components/fleet_package/contexts/advanced_fields_http_context.tsx

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { createContext, useMemo, useState } from 'react';
+import {
+  IHTTPAdvancedFields,
+  ConfigKeys,
+  Mode,
+  ResponseBodyIndexPolicy,
+  HTTPMethod,
+} from '../types';
+
+interface IHTTPAdvancedFieldsContext {
+  setFields: React.Dispatch<React.SetStateAction<IHTTPAdvancedFields>>;
+  fields: IHTTPAdvancedFields;
+  defaultValues: IHTTPAdvancedFields;
+}
+
+interface IHTTPAdvancedFieldsContextProvider {
+  children: React.ReactNode;
+  defaultValues?: IHTTPAdvancedFields;
+}
+
+export const initialValues = {
+  [ConfigKeys.PASSWORD]: '',
+  [ConfigKeys.PROXY_URL]: '',
+  [ConfigKeys.RESPONSE_BODY_CHECK_NEGATIVE]: [],
+  [ConfigKeys.RESPONSE_BODY_CHECK_POSITIVE]: [],
+  [ConfigKeys.RESPONSE_BODY_INDEX]: ResponseBodyIndexPolicy.ON_ERROR,
+  [ConfigKeys.RESPONSE_HEADERS_CHECK]: {},
+  [ConfigKeys.RESPONSE_HEADERS_INDEX]: true,
+  [ConfigKeys.RESPONSE_STATUS_CHECK]: [],
+  [ConfigKeys.REQUEST_BODY_CHECK]: {
+    value: '',
+    type: Mode.TEXT,
+  },
+  [ConfigKeys.REQUEST_HEADERS_CHECK]: {},
+  [ConfigKeys.REQUEST_METHOD_CHECK]: HTTPMethod.GET,
+  [ConfigKeys.USERNAME]: '',
+};
+
+export const defaultContext: IHTTPAdvancedFieldsContext = {
+  setFields: (_fields: React.SetStateAction<IHTTPAdvancedFields>) => {
+    throw new Error('setFields was not initialized, set it when you invoke the context');
+  },
+  fields: initialValues,
+  defaultValues: initialValues,
+};
+
+export const HTTPAdvancedFieldsContext = createContext(defaultContext);
+
+export const HTTPAdvancedFieldsContextProvider = ({
+  children,
+  defaultValues = initialValues,
+}: IHTTPAdvancedFieldsContextProvider) => {
+  const [fields, setFields] = useState<IHTTPAdvancedFields>(defaultValues);
+
+  const value = useMemo(() => {
+    return { fields, setFields, defaultValues };
+  }, [fields, defaultValues]);
+
+  return <HTTPAdvancedFieldsContext.Provider value={value} children={children} />;
+};
```

Comment:
at the end you can export context as a hook, then you will not have to do useContext while consuming context

```
export const useHTTPAdvancedFieldsContext = () => useContext(HTTPAdvancedFieldsContext);
```

Reply:
if you want to further optimise context usage, i guess you can create even more context for state and dispatch actions 
though will have to see how much value it will provide.
https://kentcdodds.com/blog/how-to-optimize-your-context-value

Reply:
this is the example

```
import React from 'react'

const CountStateContext = React.createContext()
const CountUpdaterContext = React.createContext()

function CountProvider(props) {
  const [count, setCount] = React.useState(0)
  return (
    <CountStateContext.Provider value={count}>
      <CountUpdaterContext.Provider value={setCount}>
        {props.children}
      </CountUpdaterContext.Provider>
    </CountStateContext.Provider>
  )
}

function useCountState() {
  const countState = React.useContext(CountStateContext)
  if (typeof countState === 'undefined') {
    throw new Error('useCountState must be used within a CountProvider')
  }
  return countState
}

function useCountUpdater() {
  const setCount = React.useContext(CountUpdaterContext)
  if (typeof setCount === 'undefined') {
    throw new Error('useCountUpdater must be used within a CountProvider')
  }
  const increment = React.useCallback(() => setCount((c) => c + 1), [setCount])
  return increment
}

export {CountProvider, useCountState, useCountUpdater}
```